File: src/lib/catalogMutations.js
import { supabase } from "./supabase";
import { canWriteAdmin, isAdminFull } from "./roles";

function normalizeError(error) {
  if (!error) return null;
  return error.message || String(error);
}

/**
 * @param {{ userId?: string, profile?: object|null }} actor
 */
export function buildProductActor(actor = {}) {
  const profile = actor.profile ?? null;
  const userId = actor.userId ?? null;
  return {
    userId,
    isAdmin: isAdminFull(profile),
    canWriteAdmin: canWriteAdmin(profile),
  };
}

const SELLER_EDITABLE_FIELDS = [
  "name_fr",
  "name_en",
  "description_fr",
  "prix",
  "stock",
  "categorie",
  "category_id",
  "ville",
  "image",
  "image_urls",
  "escrow",
];

/**
 * Met à jour un produit (vendeur = champs limités + vendeur_id ; admin = tous champs passés).
 */
export async function updateProduct({ productId, payload, actor }) {
  if (!productId) return { ok: false, error: "ID produit manquant" };

  const { userId, isAdmin } = buildProductActor(actor);
  const clean = { ...payload, updated_at: new Date().toISOString() };

  if (!isAdmin) {
    if (!userId) return { ok: false, error: "Non authentifié" };
    const filtered = {};
    for (const key of SELLER_EDITABLE_FIELDS) {
      if (clean[key] !== undefined) filtered[key] = clean[key];
    }
    const { data, error } = await supabase
      .from("products")
      .update(filtered)
      .eq("id", productId)
      .eq("vendeur_id", userId)
      .select()
      .maybeSingle();

    if (error) return { ok: false, error: normalizeError(error) };
    if (!data) return { ok: false, error: "Produit introuvable ou accès refusé" };
    return { ok: true, data };
  }

  const { data, error } = await supabase
    .from("products")
    .update(clean)
    .eq("id", productId)
    .select()
    .maybeSingle();

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, data };
}

/**
 * Active / désactive un produit.
 */
export async function toggleProductActive({ productId, currentActive, actor }) {
  return updateProduct({
    productId,
    payload: { actif: !currentActive },
    actor,
  });
}

/**
 * Supprime un produit via RPC (soft si commandes, hard sinon).
 * @param {{ hardDelete?: boolean }} opts — hardDelete réservé admin sans commandes
 */
export async function deleteProduct({ productId, actor, hardDelete = false }) {
  if (!productId) return { ok: false, error: "ID produit manquant" };

  const { isAdmin, userId } = buildProductActor(actor);
  if (!isAdmin && !userId) return { ok: false, error: "Non authentifié" };

  const { data, error } = await supabase.rpc("fn_delete_product", {
    p_product_id: productId,
    p_hard_delete: Boolean(hardDelete && isAdmin),
  });

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, mode: data?.mode || "unknown", hasOrders: data?.has_orders === true, data };
}

/**
 * Vérifie si un produit a des commandes (via RPC).
 */
export async function productHasOrders(productId) {
  const { data, error } = await supabase.rpc("fn_product_has_orders", {
    p_product_id: productId,
  });
  if (error) return { ok: false, error: normalizeError(error), hasOrders: false };
  return { ok: true, hasOrders: Boolean(data) };
}

File: mobile/lib/services/catalog_mutations.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../models/user_profile.dart';

class CatalogMutationResult {
  const CatalogMutationResult({
    required this.ok,
    this.error,
    this.mode,
    this.hasOrders = false,
    this.data,
  });

  final bool ok;
  final String? error;
  final String? mode;
  final bool hasOrders;
  final Map<String, dynamic>? data;
}

/// Mutations produits — aligné sur src/lib/catalogMutations.js (web).
class CatalogMutations {
  CatalogMutations(this._client);

  final SupabaseClient _client;

  static const _sellerFields = {
    'name_fr',
    'name_en',
    'description_fr',
    'prix',
    'stock',
    'categorie',
    'category_id',
    'ville',
    'image',
    'image_urls',
    'escrow',
  };

  bool _isAdmin(UserProfile? profile) =>
      profile?.role == 'admin' || profile?.role == 'superadmin';

  Future<CatalogMutationResult> updateProduct({
    required String productId,
    required Map<String, dynamic> payload,
    String? userId,
    UserProfile? profile,
  }) async {
    final admin = _isAdmin(profile);
    final clean = Map<String, dynamic>.from(payload);
    clean['updated_at'] = DateTime.now().toUtc().toIso8601String();

    if (!admin) {
      if (userId == null) {
        return const CatalogMutationResult(ok: false, error: 'Non authentifié');
      }
      final filtered = <String, dynamic>{};
      for (final key in _sellerFields) {
        if (clean.containsKey(key)) filtered[key] = clean[key];
      }
      try {
        final row = await _client
            .from('products')
            .update(filtered)
            .eq('id', productId)
            .eq('vendeur_id', userId)
            .select()
            .maybeSingle();
        if (row == null) {
          return const CatalogMutationResult(
            ok: false,
            error: 'Produit introuvable ou accès refusé',
          );
        }
        return CatalogMutationResult(
          ok: true,
          data: Map<String, dynamic>.from(row),
        );
      } on PostgrestException catch (e) {
        return CatalogMutationResult(ok: false, error: e.message);
      }
    }

    try {
      final row = await _client
          .from('products')
          .update(clean)
          .eq('id', productId)
          .select()
          .maybeSingle();
      return CatalogMutationResult(
        ok: true,
        data: row != null ? Map<String, dynamic>.from(row) : null,
      );
    } on PostgrestException catch (e) {
      return CatalogMutationResult(ok: false, error: e.message);
    }
  }

  Future<CatalogMutationResult> toggleProductActive({
    required String productId,
    required bool currentActive,
    String? userId,
    UserProfile? profile,
  }) {
    return updateProduct(
      productId: productId,
      payload: {'actif': !currentActive},
      userId: userId,
      profile: profile,
    );
  }

  Future<CatalogMutationResult> deleteProduct({
    required String productId,
    String? userId,
    UserProfile? profile,
    bool hardDelete = false,
  }) async {
    if (userId == null && !_isAdmin(profile)) {
      return const CatalogMutationResult(ok: false, error: 'Non authentifié');
    }
    try {
      final data = await _client.rpc(
        'fn_delete_product',
        params: {
          'p_product_id': productId,
          'p_hard_delete': hardDelete && _isAdmin(profile),
        },
      );
      final map = data is Map ? Map<String, dynamic>.from(data) : <String, dynamic>{};
      return CatalogMutationResult(
        ok: true,
        mode: map['mode']?.toString(),
        hasOrders: map['has_orders'] == true,
      );
    } on PostgrestException catch (e) {
      return CatalogMutationResult(ok: false, error: e.message);
    }
  }

  Future<bool> productHasOrders(String productId) async {
    try {
      final data = await _client.rpc(
        'fn_product_has_orders',
        params: {'p_product_id': productId},
      );
      return data == true;
    } catch (_) {
      return false;
    }
  }
}
